### DISCUSSION TITLE
tox-backticks replacement: tox-backtocks

### DISCUSSION THREAD

**Discussion by Anvil** [tox-dev/tox#2938] [Show and tell]
Dear Tox maintainers, Dear Tox Community.

It is my pleasure to announce that I'm currently developing a replacement plugin for `tox-backticks`.
Its name is [tox-backtocks](https://github.com/Anvil/tox-backtocks)

The code is not really mature right now, but it *is* working for basic cases.

My plan is to make it work with the projects we currently have in the place i work at, in order to finally be able to upgrade to tox version 4. If I can manage to satisfy some or all of your use cases, I would be delighted. If backtocks fails for you, please feel free to report said failure.

My plan is also make things easier for users. Like:
* do not interfere with char escaping process
* match shells behaviour by removing trailing newlines characters.

Feel free to tell me if any thing really bothered you before with backticks, or is bothering you right now with backtocks, or if you want me to consider some specific aspect/feature in order not to break it :clown_face: 

Also, I am learning tox internals: what they allow you to do and what they cannot seem to allow to do... If you read tox-backtocks source code, you will see that I had to make use of some tox objects private attributes. I am happy I could made it work, but I'm not happy for what it implies in terms of fragility. 

In that regard, I am very open to suggestions:
* Is there a better way to do it?
* If not, would  it be possible to make the `SetEnv` class evolve in order to expose variables values along with their name? (It is the purpose of the `tox-backtocks`, after all, to actually *change* the way values are computed)
* If so, then should I provide such a patch?

(I dont take myself seriously... why so serious heh ? but I consider my code quality seriously)

Best regards. Take care.

**Comment by gaborbernat** [comment]
It's not clear what this plugin does and why one would one to use it 🤔

**Reply by Anvil** [reply]
Oh, I'm sorry, I thought people actually knew about backquotes and tox-backticks. My bad.

So this plugin allows you to use backquote expressions in the `set_env` section.

Basically instead of having `passenv = some_var` in the tox.ini and doing that:

```bash
#!/bin/bash
# Do some stuff
export some_var=`head -1 /etc/passwd`
tox -e asdf
```

tox-backtocks allows you to directly write this:
```ini
[testenv:asdf]
setenv =
    some_var = `head -1 /etc/passwd`
    some_other_var = `vault some_secret`
    something = `basically run a command`
```

Like one would do in bourne shell syntax. 

Does this make sense ?

**Reply by gaborbernat** [reply]
It does and explains why I never heard about this or used it. It goes against the spirit of tox, the cross platform part of it to be specific. So I personally would not recommend it, but it's an interesting case study nevertheless.

**Reply by Anvil** [reply]
I understand the portability benefits of cross-platform, and thank you very much for having that in mind. 
In the same time, there is a need to perform such actions, in a quite painless way. 

How do you think we could achieve such feature without having plugins like backticks/backtocks ?

I don't have any attachment to this plugin. If I can get rid of it, I will :)

**Reply by Anvil** [reply]
Also, bash exists on nearly every platform (Linux, Windows, Mac, ...). I rely on bash to evaluate the backquote expressions. Does this match tox cross-platform criteria or did you have anything bigger than that in mind ?

Or do you have more concerns about the non-portability of the command itself ? (which could be dealt with the same way a python module would be unix-only or windows-only: let the user handle those case)

**Reply by gaborbernat** [reply]
While it might exist it's not available out of the box. I'd likely use just python 😅 rather than bash here 🤔 but please ignore me 😎
